### DISCUSSION TITLE
Rule Proposal: Disallow reading @Input() in constructor

### DISCUSSION THREAD

**Discussion by Friendseeker** [angular-eslint/angular-eslint#1321] [New Rule Proposals]
For Example
```typescript
@Component
export class Test {
   @Input() input: boolean;
   constructor() {
       something(input);
   }
}
```
Should be flagged as `input` is `undefined` in constructor.

Similarly, reading `ViewChild()` in constructor and certain lifecycle methods should be flagged.

**Comment by SamuelT-Beslogic** [comment]
With `"strictPropertyInitialization": true` in tsconfig you should get:
`Property 'input' has no initializer and is not definitely assigned in the constructor. ts(2564)`

For use with `@Input({ required: true })` you can either assign a default value: `@Input({ required: true }) input = false` or use the null-assertion operator: `@Input({ required: true }) input!: boolean`. The second option isn't type-safe, but I use a custom `no-restricted-syntax` ESLint rule config that ensures it's only used with properties that are required inputs:
```json
"no-restricted-syntax": [
  "error",
  {
    "selector":
      "PropertyDefinition[definite=true]:not(:has(Decorator[expression.callee.name=\"Input\"] Property[key.name=\"required\"][value.value=true]))",
    "message":
      "Forbidden non-null property assertion. https://typescript-eslint.io/rules/no-non-null-assertion . This is only allowed on required component @Input({ required:true }). https://angular.io/api/core/Input#required"
  }
]
```
